### DISCUSSION TITLE
How to get mock outputs in the case of Map scenario

### DISCUSSION THREAD

**Discussion by s-u-b-h-a-k-a-r** [gruntwork-io/knowledge-base#677] [Help]
Please help how to get the mock outputs to refer as dependency in the below case

Subnets response
subnets = {
  "us-west1/te-admin-subnet" = {
    "creation_timestamp" = "2023-03-07T23:43:13.435-08:00"
    "description" = ""
    "external_ipv6_prefix" = ""
    "fingerprint" = tostring(null)
    "gateway_address" = "10.0.0.1"
    "id" = "projects/systems/regions/us-west1/subnetworks/te-admin-subnet"
    "ip_cidr_range" = "10.0.0.0/16"
    "ipv6_access_type" = ""
    "ipv6_cidr_range" = ""
    "log_config" = tolist([])
    "name" = "te-admin-subnet"
    "network" = "https://www.googleapis.com/compute/v1/projects/systems/global/networks/te-admin-vpc"
    "private_ip_google_access" = true
    "private_ipv6_google_access" = "DISABLE_GOOGLE_ACCESS"
    "project" = "systems"
    "purpose" = "PRIVATE"
    "region" = "us-west1"
    "role" = ""
    "secondary_ip_range" = tolist([
      {
        "ip_cidr_range" = "10.10.0.0/16"
        "range_name" = "te-admin-ip-range-pods"
      },
      {
        "ip_cidr_range" = "10.100.0.0/16"
        "range_name" = "te-admin-ip-range-services"
      },
    ])
    "self_link" = "https://www.googleapis.com/compute/v1/projects/systems/regions/us-west1/subnetworks/te-admin-subnet"
    "stack_type" = "IPV4_ONLY"
    "timeouts" = null /* object */
  }
}

terraform {
  source = "tfr:///terraform-google-modules/network/google//modules/subnets?version=6.0.1"
}

dependency "subnets" {
  config_path = "../subnets"
  mock_outputs = **{subnets = {dummy = {}}}**
  }

inputs = {
  network                       = dependency.vpc.outputs.network_name
  subnetwork                    = **dependency.subnets.outputs.subnets["dummy"].name**
  ip_range_pods                 = "${local.name}-${local.env}-ip-range-pods"
  ip_range_services             = "${local.name}-${local.env}-ip-range-services"
  master_ipv4_cidr_block  = "10.100.100.0/28"
}

---

<ins datetime="2023-03-08T09:43:00Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109960">Tracked in ticket #109960</a></p>
</ins>

**Comment by gitsstewart** [comment]
@s-u-b-h-a-k-a-r is this question in reference to one of our open source projects or is this a general Terraform question?

**Reply by s-u-b-h-a-k-a-r** [reply]
Hi @gitsstewart its more related to terragrunt how mock outputs should be used.

**Comment by denis256** [comment]
Hi,
I think should be defined `name` field in mock object to be referenced through ".":
```
dependency "subnets" {
...
  mock_outputs = {
    subnets   = {
      dummy = {
        name = "test"
      }
    }
  }
....
}

inputs = {
  subnetwork = dependency.subnets.outputs.subnets["dummy"].name
}
```
or referenced through `["fieldName"]`:

```
inputs = {
  subnetwork = dependency.module2.outputs.subnets["dummy"]["name"]
}

```

**Reply by s-u-b-h-a-k-a-r** [reply]
> Hi, I think should be defined `name` field in mock object to be referenced through ".":
> 
> ```
> dependency "subnets" {
> ...
>   mock_outputs = {
>     subnets   = {
>       dummy = {
>         name = "test"
>       }
>     }
>   }
> ....
> }
> 
> inputs = {
>   subnetwork = dependency.subnets.outputs.subnets["dummy"].name
> }
> ```
> 
> or referenced through `["fieldName"]`:
> 
> ```
> inputs = {
>   subnetwork = dependency.module2.outputs.subnets["dummy"]["name"]
> }
> ```
Thank you
Tried both but getting the same error.tried both the versions
69-78: Invalid index; The given key does not identify an element in this collection value., and 1 other diagnostic(s)
subnetwork                    = dependency.subnets.outputs.subnets["dummy"].name
subnetwork                    = dependency.subnets.outputs.subnets["dummy"]["name"]

```subnets = {
  "us-west1/te-admin-subnet" = {
    "ip_cidr_range" = "10.0.0.0/16"
     "name" = "te-admin-subnet"
  }
}```

I am referencing like this please correct me
```dependency "subnets" {
  config_path = "../subnets"
  mock_outputs = {
    subnets   = {
      dummy = {
        name = "test"
        ip_cidr_range = "0.0.0.0/0"
      }
    }
  }
}```
```inputs = {
  network                       = dependency.vpc.outputs.network_name
  subnetwork                    = dependency.subnets.outputs.subnets["dummy"].name
  ip_range_pods                 = "${local.name}-${local.env}-ip-range-pods"
  ip_range_services             = "${local.name}-${local.env}-ip-range-services"
  master_ipv4_cidr_block  = "10.100.100.0/28"
  master_authorized_networks = [
    {
      cidr_block   =dependency.subnets.outputs.subnets["dummy"].ip_cidr_range
      display_name = "VPC"
    }
  ]
  
}```

**Reply by denis256** [reply]
Hi, 
I was wondering if `mock_outputs_allowed_terraform_commands` is defined to specify when mock data should be used

**Reply by s-u-b-h-a-k-a-r** [reply]
> Hi, I was wondering if `mock_outputs_allowed_terraform_commands` is defined to specify when mock data should be used

Yes used but still the same error. 
```
dependency "vpc" {
  config_path  = "../../vpc"
  mock_outputs = {
    network_name = "dummy-id"
  }
  mock_outputs_allowed_terraform_commands = ["apply", "plan", "destroy", "output"]
}

dependency "subnets" {
  config_path  = "../subnets"
  mock_outputs = {
    subnets   = {
      dummy = {
        name = "test"
        ip_cidr_range = "0.0.0.0/0"
      }
    }
  }
  mock_outputs_allowed_terraform_commands = ["apply", "plan", "destroy", "output"]
}
```
